docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,435 @@
+So identify which actual positions in the array are being gotten/set is common work to both primal and gradient computations.
+We really don't want to deal with fancy indexing types during the pullback, because there are weird edge cases like indexing in such a way that the same element is output twice (and thus we have 2 sensitivities we need to add to it).
+We can pull the `to_indices` out of the primal computation and remember the plain indexes used, then can reuse them to set gradients during the pullback.
+
+See the [code for this in ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl/blob/v0.7.49/src/rulesets/Base/indexing.jl)
+
+### `exp(::Matrix)`
+[Matrix Functions](https://en.wikipedia.org/wiki/Matrix_function) are functions that are generalized to operate on matrixes rather than scalars.
+Note that this is distinct from simply element-wise application of the function to the matrix's elements.
+[Matrix Exponential](https://en.wikipedia.org/wiki/Matrix_exponential) `exp(::Matrix)` is one particularly important matrix function.
+
+
+[Al-Mohy, Awad H. and Higham, Nicholas J. (2009) _Computing the Fréchet Derivative of the Matrix Exponential, with an application to Condition Number Estimation_. SIAM Journal On Matrix Analysis and Applications., 30 (4). pp. 1639-1657. ISSN 1095-7162](http://eprints.maths.manchester.ac.uk/1218/), published a method for this.
+It is pretty complex and very cool.
+To quote its abstract (emphasis mine):
+
+> The algorithm is derived from the scaling and squaring method by differentiating the Padé approximants and the squaring recurrence, **re-using quantities computed during the evaluation of the Padé approximant**, and intertwining the recurrences in the squaring phase.
+
+Julia does in fact use a Padé approximation to compute `exp(::Matrix)`.
+So we can extract the code for that into our augmented primal, and add remembering the intermediate quantities that are to be used.
+See the [code for this in ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl/blob/v0.7.49/src/rulesets/LinearAlgebra/matfun.jl)
+
+
+An interesting thing here that may be of concern to some:
+if Julia changes the algorithm it it uses to compute `exp(::Matrix)` then during an AD primal pass, it will continue to use the old Padé approximation based algorithm.
+This is a real thing that might happen, there are many other algorithms that can compute the matrix exponential.
+Further, perhaps there might be an improvement to the exact coefficient or cut-offs used by julia's current Padé approximation.
+This is not a breaking change on Julia's behalf: [exact floating point numerical values are not generally considered part of the SemVer-bound API](http://colprac.sciml.ai/#changes-that-are-not-considered-breaking).
+Rather only the general accuracy of the computed value relative to the true mathematical value (e.g. for common scalar operations Julia promises 1ULP).
```

Comment:
Can you elaborate on what 1ULP means?

Reply:
I really wish I could.
I actually don't know, just that it is how we measure accuracy.
But I can cross link to wikipedia.

```suggestion
Rather only the general accuracy of the computed value relative to the true mathematical value (e.g. for common scalar operations Julia promises 1 [ULP](https://en.wikipedia.org/wiki/Unit_in_the_last_place)).
```